c++/src/kj/async-coroutine-test.c++

```diff
@@ -143,13 +143,82 @@ KJ_TEST("Coroutines can be canceled while suspended") {
     auto neverDone = kj::Promise<int>(kj::NEVER_DONE);
     neverDone = neverDone.attach(kj::heap<Counter>(count));
     auto promise = coro(kj::mv(neverDone));
-    promise.poll(waitScope);
+    KJ_EXPECT(!promise.poll(waitScope));
   }
 
   // Stack variables on both sides of a co_await, plus coroutine arguments are destroyed.
   KJ_EXPECT(count == 3);
 }
 
+kj::Promise<void> deferredThrowCoroutine(kj::Promise<void> awaitMe) {
+  KJ_DEFER(kj::throwFatalException(KJ_EXCEPTION(FAILED, "thrown during unwind")));
+  co_await awaitMe;
+  co_return;
+};
+
+KJ_TEST("Exceptions during suspended coroutine frame-unwind propagate via destructor") {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  auto exception = KJ_ASSERT_NONNULL(kj::runCatchingExceptions([&]() {
+    deferredThrowCoroutine(kj::NEVER_DONE);
+  }));
+
+  KJ_EXPECT(exception.getDescription() == "thrown during unwind");
+};
+
+KJ_TEST("Exceptions during suspended coroutine frame-unwind do not cause a memory leak") {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  // We can't easily test for memory leaks without hooking operator new and delete. However, we can
+  // arrange for the test to crash on failure, by having the coroutine suspend at a promise that we
+  // later fulfill, thus arming the Coroutine's Event. If we destroy the coroutine in this state,
```

Comment:
"If we *fail* to destroy the coroutine ..." I mean.

Reply:
https://github.com/capnproto/capnproto/pull/1104/commits/c3e2dcec30de0f07500ed49faee855d137e0dfbf